File: src/main/kotlin/me/snwy/oasis/Environment.kt
package me.snwy.oasis

import java.util.*

class Environment(private val enclosing: Environment? = null) {
    @JvmField
    var values: MutableMap<Int, Any?> = HashMap()

    fun define(name: Int, value: Any?) {
        if (values.containsKey(name)) {
            throw RuntimeError(line, "Variable with this name already defined")
        }
        values[name] = value
    }

    fun define(name: String, value: Any?) {
        if (values.containsKey(name.hashCode())) {
            throw RuntimeError(line, "Variable with this name already defined")
        }
        values[name.hashCode()] = value
    }

    fun get(name: Int): Any? {
        if (name in values) {
            return values[name]
        }
        if (enclosing != null) {
            return enclosing.get(name)
        }
        throw RuntimeError(line, "Undefined variable '${Optimizer.nameMap[name]}'")
    }

    fun assign(name: Int, value: Any?) {
        if (name in Optimizer.immutables) {
            throw RuntimeError(line, "Cannot assign to immutable variable")
        }
        if (values.containsKey(name)) {
            values[name] = value
            return
        }
        if (enclosing != null) {
            enclosing.assign(name, value)
            return
        }
        throw RuntimeError(line, "Undefined variable '${Optimizer.nameMap[name]}'")
    }

    private fun assign(name: String, value: Any?) {
        if (name.hashCode() in Optimizer.immutables) {
            throw RuntimeError(line, "Cannot assign to immutable variable")
        }
        if (values.containsKey(name.hashCode())) {
            values[name.hashCode()] = value
            return
        }
        if (enclosing != null) {
            enclosing.assign(name, value)
            return
        }
        throw RuntimeError(line, "Undefined variable '${name}'")
    }

    override fun toString(): String {
        return values.toString()
    }
}

File: core/environment.go
package core

import "fmt"

type OasisEnvironment struct {
	Values map[string]any
	Consts []string
}

func NewEnvironment() OasisEnvironment {
	return OasisEnvironment{
		Values: map[string]any{},
		Consts: []string{},
	}
}

func (e *OasisEnvironment) Define(name string, isConst bool, value any) error {
	if _, ok := e.Values[name]; ok {
		return fmt.Errorf("Variable %s already defined", name)
	}
	e.Values[name] = value
	if isConst {
		e.Consts = append(e.Consts, name)
	}
	return nil
}

func (e *OasisEnvironment) Assign(name string, value any) error {
	if _, ok := e.Values[name]; !ok {
		return fmt.Errorf("Variable %s not defined", name)
	}
	e.Values[name] = value
	return nil
}

func (e *OasisEnvironment) Get(name string) (any, error) {
	if _, ok := e.Values[name]; !ok {
		return nil, fmt.Errorf("Variable %s not defined", name)
	}
	return e.Values[name], nil
}
